src/source/native.rs

```diff
@@ -70,42 +69,59 @@ impl TCPStreamHandler for NativeStreamHandler {
         poller: &mio::Poll,
         stream: mio::net::TcpStream,
     ) -> () {
-        let mut reader = io::BufReader::new(stream);
-        loop {
+        let mut streaming = true;
+        let mut reader = BufferedPayload::new(stream.try_clone().unwrap());
+        while streaming {
             let mut events = mio::Events::with_capacity(1024);
             match poller.poll(&mut events, None) {
                 Err(e) => panic!(format!("Failed during poll {:?}", e)),
                 Ok(_num_events) => for event in events {
                     match event.token() {
                         constants::SYSTEM => return,
                         _stream_token => {
-                            let rchans = chans.clone();
-                            self.handle_stream_payload(rchans, tags, &mut reader);
+                            while streaming {
+                                match reader.read() {
+                                    Ok(mut raw) => {
+                                        let rchans = chans.clone();
+                                        self.handle_stream_payload(rchans, tags, &mut raw);
+                                    }
+                                    Err(PayloadErr::WouldBlock) => {
+                                        //Not enough data yet.  Try again.
+                                        break;
+                                    }
+                                    Err(PayloadErr::IO(ref e)) if e.kind() == ErrorKind::UnexpectedEof => {
+                                        //Client went away.  Shut it down (gracefully).
```

Comment:
Telemetry?

Reply:
None existed before.

Perhaps we pay down that tech debt in a separate pr?

Reply:
It's expected that clients will come and go but I'm not opposed to telemetry around this.